bitcoin/test/fuzz/miniscript_random.cpp

```diff
@@ -0,0 +1,417 @@
+        return LookupHash(hash, preimage, test_data->hash256_preimages);
+    }
+    miniscript::Availability SatHASH160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
+        return LookupHash(hash, preimage, test_data->hash160_preimages);
+    }
+};
+
+//! Context to check a satisfaction against the pre-computed data.
+struct CheckerContext: BaseSignatureChecker {
+    TestData *test_data;
+
+    // Signature checker methods. Checks the right dummy signature is used. Always assumes timelocks are
+    // correct.
+    bool CheckECDSASignature(const std::vector<unsigned char>& sig, const std::vector<unsigned char>& vchPubKey,
+                             const CScript& scriptCode, SigVersion sigversion) const override
+    {
+        const CPubKey key{vchPubKey};
+        const auto it = test_data->dummy_sigs.find(key);
+        if (it == test_data->dummy_sigs.end()) return false;
+        return it->second == sig;
+    }
+    bool CheckLockTime(const CScriptNum& nLockTime) const override { return true; }
+    bool CheckSequence(const CScriptNum& nSequence) const override { return true; }
+};
+
+// The various contexts
+TestData TEST_DATA;
+ParserContext PARSER_CTX;
+SatisfierContext SATISFIER_CTX;
+CheckerContext CHECKER_CTX;
+// A dummy scriptsig to pass to VerifyScript (we always use Segwit v0).
+const CScript DUMMY_SCRIPTSIG;
+// We generate the pseudorandom nodes recursively, this puts a bound.
+static constexpr size_t MAX_NESTED_DEPTH = 402;
+
+using NodeType = miniscript::NodeType;
+using NodeRef = miniscript::NodeRef<CPubKey>;
+using miniscript::operator"" _mst;
+
+//! Construct a miniscript node as a shared_ptr.
+template<typename... Args> NodeRef MakeNodeRef(Args&&... args) { return miniscript::MakeNodeRef<CPubKey>(std::forward<Args>(args)...); }
+
+/**
+ * Generate a Miniscript node based on the fuzzer's input.
+ * Note this does not attempt to produce a well-typed (let alone safe) Miniscript node.
+ */
+NodeRef GenNode(FuzzedDataProvider& provider, const miniscript::Type typ, const size_t recursion_depth);
+
+/**
+ * Generate a pseudorandom node. If invalid, return NULL.
+ * Used to cut-through as a node with invalid subs will never be valid.
+ */
+NodeRef GenValidNode(FuzzedDataProvider& provider, const miniscript::Type typ, const size_t recursion_depth)
+{
+    const NodeRef node = GenNode(provider, typ, recursion_depth);
+    if (!node || !node->IsValid()) return {}; // TODO!
+    return node;
+}
+
+//! Generate a vector of miniscript nodes of the given types.
+std::vector<NodeRef> MultiNode(FuzzedDataProvider& provider, const std::vector<miniscript::Type>& types,
+                               const size_t recursion_depth)
+{
+    std::vector<NodeRef> subs;
+    for (size_t i = 0; i < types.size(); ++i) {
+        const NodeRef sub = GenValidNode(provider, types[i], recursion_depth);
```

Comment:
Don't make this `const`, and you can use `subs.push_back(std::move(sub))` below.

NodeRefs are shared_ptrs, which are cheaply, but not trivially, copyable (they need an atomic refcount increment).